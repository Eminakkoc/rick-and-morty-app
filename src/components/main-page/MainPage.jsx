import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Page from 'components/page/Page';
import CharacterCard from 'components/character-card/CharacterCard';
import LoadingCharacterCard from 'components/character-card/LoadingCharacterCard';

import history from 'common/history/History';
import paths from 'routes/paths';

import constants from 'common/constants';

import { graphql } from 'react-apollo';
import gql from 'graphql-tag';

const redirectToDetailsPage = () => {
  history.push({
    pathname: paths.DETAILS_PAGE,
  });
};

class MainPage extends Component {
  constructor(props) {
    super(props);

    this.state = {
      initialLoad: true,
    };
  }

  componentWillReceiveProps(nextProps) {
    const { initialLoad } = this.state;
    const { data: { loading } } = this.props;
    const { addCharacters, selectedCharacter } = this.props;

    if (initialLoad && !nextProps.data.loading) {
      this.setState({
        initialLoad: false,
      });
    }

    if (loading && !nextProps.data.loading && nextProps.data.characters) {
      addCharacters(nextProps.data.characters.results);
    }

    if (selectedCharacter !== nextProps.selectedCharacter) {
      redirectToDetailsPage();
    }
  }

  render() {
    const { data: { loading } } = this.props;
    const { characterList, selectCharacter } = this.props;
    const { initialLoad } = this.state;

    return (
      <Page>
        <div className="main-page-container">
          {
            loading && initialLoad ? (
              <div>
                <LoadingCharacterCard />
                <LoadingCharacterCard />
              </div>
            ) : (
              characterList.map(character => (
                <CharacterCard
                  name={character.name}
                  image={character.image}
                  detailsCallback={selectCharacter}
                />
              ))
            )
          }
        </div>
      </Page>
    );
  }
}

const query = gql`
${constants.GET_CHARACTERS}
`;

MainPage.propTypes = {
  data: PropTypes.shape({
    loading: PropTypes.bool,
    characters: PropTypes.shape({
      results: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string,
        image: PropTypes.string,
      })),
    }),
  }),
  characterList: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    image: PropTypes.string,
  })),
  addCharacters: PropTypes.func.isRequired,
  selectCharacter: PropTypes.func.isRequired,
  selectedCharacter: PropTypes.shape({
    name: PropTypes.string,
    image: PropTypes.string,
  }),
};

MainPage.defaultProps = {
  data: {
    loading: false,
    characters: {
      results: [{
        name: '',
        image: '',
      }],
    },
  },
  characterList: [{
    name: '',
    image: '',
  }],
  selectedCharacter: {
    name: '',
    image: '',
  },
};

export default graphql(query)(MainPage);
