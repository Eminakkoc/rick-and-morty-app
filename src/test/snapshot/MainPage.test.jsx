// MainPage.test.js
import React from 'react';
import MainPage from '../../components/main-page/MainPage';
import renderer from 'react-test-renderer';

test('Main Page Component Structure', () => {
  const component = renderer.create(
    <MainPage
      data={{
        refetch: () => {},
      }}
      addCharacters={() => {}}
      selectCharacter={() => {}}
    />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
