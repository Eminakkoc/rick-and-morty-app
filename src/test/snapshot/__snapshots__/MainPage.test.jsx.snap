// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Page Component Structure 1`] = `
<div
  className="page-container"
>
  <div
    className="page-header"
  >
    <div
      className="page-logo"
    />
  </div>
  <div
    className="main-page-container"
  >
    <div
      className="MuiPaper-root MuiPaper-elevation1 MuiCard-root card MuiPaper-rounded"
    >
      <button
        className="MuiButtonBase-root MuiCardActionArea-root"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <div
          className="card-wrapper"
        >
          <div
            className="MuiCardContent-root card-info"
          >
            
          </div>
          <div
            className="MuiCardMedia-root card-image"
            style={
              Object {
                "backgroundImage": "url(\\"./\\")",
              }
            }
            title="Rick And Morty Character"
          />
        </div>
        <span
          className="MuiCardActionArea-focusHighlight"
        />
      </button>
    </div>
  </div>
</div>
`;
